includes/Framework/PaymentGateway/Integrations/Payment_Gateway_Integration_Pre_Orders.php

```diff
@@ -87,26 +87,33 @@ public function add_support() {
 	 * @return boolean true if tokenization should be forced, false otherwise
 	 */
 	public function maybe_force_tokenization( $force_tokenization ) {
+		Performance_Logger::start( 'force_tokenization', $this->get_gateway()->get_plugin() );
 
-		// pay page with pre-order?
-		$pay_page_pre_order = false;
-		if ( $this->get_gateway()->is_pay_page_gateway() ) {
+		try {
+			// pay page with pre-order?
+			$pay_page_pre_order = false;
+			if ( $this->get_gateway()->is_pay_page_gateway() ) {
 
-			$order_id = $this->get_gateway()->get_checkout_pay_page_order_id();
+				$order_id = $this->get_gateway()->get_checkout_pay_page_order_id();
 
-			if ( $order_id ) {
-				$pay_page_pre_order = class_exists( 'WC_Pre_Orders_Order' ) && class_exists( 'WC_Pre_Orders_Product' ) && \WC_Pre_Orders_Order::order_contains_pre_order( $order_id ) && \WC_Pre_Orders_Product::product_is_charged_upon_release( \WC_Pre_Orders_Order::get_pre_order_product( $order_id ) );
+				if ( $order_id ) {
+					$pay_page_pre_order = class_exists( 'WC_Pre_Orders_Order' ) && class_exists( 'WC_Pre_Orders_Product' ) && \WC_Pre_Orders_Order::order_contains_pre_order( $order_id ) && \WC_Pre_Orders_Product::product_is_charged_upon_release( \WC_Pre_Orders_Order::get_pre_order_product( $order_id ) );
+				}
 			}
-		}
 
-		if ( ( class_exists( 'WC_Pre_Orders_Cart' ) && class_exists( 'WC_Pre_Orders_Product' ) && \WC_Pre_Orders_Cart::cart_contains_pre_order() && \WC_Pre_Orders_Product::product_is_charged_upon_release( \WC_Pre_Orders_Cart::get_pre_order_product() ) ) ||
-			$pay_page_pre_order ) {
+			if ( ( class_exists( 'WC_Pre_Orders_Cart' ) && class_exists( 'WC_Pre_Orders_Product' ) && \WC_Pre_Orders_Cart::cart_contains_pre_order() && \WC_Pre_Orders_Product::product_is_charged_upon_release( \WC_Pre_Orders_Cart::get_pre_order_product() ) ) ||
+				$pay_page_pre_order ) {
 
-			// always tokenize the card for pre-orders that are charged upon release
-			$force_tokenization = true;
-		}
+				// always tokenize the card for pre-orders that are charged upon release
+				$force_tokenization = true;
+			}
 
-		return $force_tokenization;
+			Performance_Logger::end( 'force_tokenization', $this->get_gateway()->get_plugin() );
+			return $force_tokenization;
+		} catch ( \Exception $e ) {
+			Performance_Logger::end( 'force_tokenization', $this->get_gateway()->get_plugin(), true );
+			throw $e;
```

Comment:
This method now throws an exception whereas previously it would only return `true` or `false`. We need to update the docblock here mentioning the new exception but we also need to ensure any places that call `maybe_force_tokenization` are properly handling the exception that could now get thrown.

Reply:
Thanks for the feedback. Instead of throwing an exception, I think we should return `false` so that we don't need to update the codebase for existing functionality, and it will work as expected.

Reply:
Updated in https://github.com/woocommerce/woocommerce-square/pull/279/commits/8105e737f717c72f47d7c40b95dd984af5959c5a